_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+For time-based log retention, use the milliseconds configuration, which has priority over the related minutes and hours configuration.
+The milliseconds configuration also updates dynamically.
+
+```properties
+# ...
+log.retention.ms=1680000
+# ...
+```
+
+The retention period is based on the time messages were appended to the segment.
+If  `log.retention.ms` is set to -1, no time limit is applied to log retention, so all logs are retained.  
+Disk usage should always be monitored, but the -1 setting is not generally recommended as it can lead to issues with full disks, which can be hard to rectify.
+
+For size-based log retention, set a maximum log size in bytes:
+
+```properties
+# ...
+log.retention.bytes=1073741824
+# ...
+```
+
+The maximum log size is for all segments in the log.
+When the maximum log size is reached, older segments are removed.
+
+Setting a maximum log size does not consider the time messages were appended to a segment.
+If that's a potential issue, you can use time-based and size-based retention. When you use both, the first threshold reached triggers the cleanup.
+
+### Removing log data with cleanup policies
+
+Kafka's log cleaner does what you expect.
+It removes old data from the log.
+Enabled by default (`log.cleaner.enable=true`), you can put some control on how the cleaner operates by defining a _cleanup policy_.
+
+```properties
+# ...
+log.cleanup.policy=compact,delete
+# ...
+```
+
+* **Delete policy**
+`delete` policy corresponds to managing logs with data retention policies. It's suitable when data does not need to be retained forever. Older segments are deleted based on log retention limits. Otherwise, if the log cleaner is not enabled, and there are no log retention limits, the log will continue to grow.
+
+* **Compact policy**
+`compact` policy guarantees to keep the most recent message for each message key. Log compaction is suitable when message values are changeable, and you want to retain the latest update. New messages are appended to the  _head_ of the log, which acts in the same way as a non-compacted log with writes appended in order. The _tail_ of a compacted log has the older messages that are deleted or compacted according to policy. Consequently, the tail has non-contiguous offsets.
+
+**Log showing key value writes with offset positions before compaction**
+
+![Image of compaction showing key value writes](/assets/images/posts/2021-05-06-broker-tuning-compaction-1.png)
+
+If you're not using keys, you can't use compaction as keys are needed to identify related messages. The latest message (with the highest offset) is kept and older messages with the same key are discarded. You can restore a message back to a previous state. Records retain their original offsets after cleanup. When consuming an offset that's no longer available in the tail, the record with the next higher offset is found.
```

Comment:
IDK what "You can restore a message back to a previous state." is supposed to mean

```suggestion
If you're not using keys, you can't use compaction as keys are needed to identify related messages. During log cleaning the latest message (with the highest offset) is kept and older messages with the same key are discarded.  Records retain their original offsets after cleanup. When consuming an offset that's no longer available in the tail, the record with the next higher offset is found.
```